ion-schema/src/isl/ranges.rs

```diff
@@ -0,0 +1,471 @@
+            } else {
+                invalid_schema_error(format!("Invalid value for range: {element}"))
+            }
+        }
+    }
+
+    /// Checks whether the given value is contained within this range.
+    pub fn contains<V: Into<T> + Clone>(&self, value: &V) -> bool {
+        self.lower.is_below(value) && self.upper.is_above(value)
+    }
+
+    fn is_bounded_and_non_empty(start: &Limit<T>, end: &Limit<T>) -> bool {
+        match (start, end) {
+            (Limit::Unbounded, Limit::Unbounded) => false,
+            (Limit::Unbounded, _) => true,
+            (_, Limit::Unbounded) => true,
+            (Limit::Closed(lower), Limit::Closed(upper)) => lower <= upper,
+            (Limit::Open(lower), Limit::Open(upper))
+            | (Limit::Open(lower), Limit::Closed(upper))
+            | (Limit::Closed(lower), Limit::Open(upper)) => lower < upper,
+        }
+    }
+}
+
+fn read_range_bound<T, F: Fn(&Element) -> Option<T>>(
+    element: &Element,
+    boundary_element: &Element,
+    unbounded_text: &str,
+    value_fn: F,
+) -> IonSchemaResult<Limit<T>> {
+    let limit = if boundary_element.as_symbol() == Some(&ion_rs::Symbol::from(unbounded_text)) {
+        isl_require!(boundary_element.annotations().is_empty() => "'{unbounded_text}' may not have annotations: {element}")?;
+        Limit::Unbounded
+    } else {
+        let upper_value: T = value_fn(boundary_element).ok_or_else(|| {
+            invalid_schema_error_raw(format!("Invalid value for range boundary: {element}"))
+        })?;
+        if boundary_element.annotations().contains("exclusive") {
+            isl_require!(boundary_element.annotations().len() == 1 => "Invalid annotation(s) on range boundary {element}")?;
+            Limit::Open(upper_value)
+        } else {
+            isl_require!(boundary_element.annotations().is_empty() => "Invalid annotation(s) on range boundary {element}")?;
+            Limit::Closed(upper_value)
+        }
+    };
+    Ok(limit)
+}
+
+impl<T: Clone + PartialEq + PartialOrd + RangeOverType> From<T> for Range<T> {
+    fn from(value: T) -> Self {
+        Range::new_single_value(value)
+    }
+}
+
+pub type UsizeRange = Range<usize>;
+impl UsizeRange {
+    /// Constructs an inclusive range from 0 to 1.
+    pub fn zero_or_one() -> Self {
+        Self::new_inclusive(0, 1).expect("This is safe to unwrap because 0 <= 1")
+    }
+
+    pub fn inclusive_endpoints(&self) -> (usize, usize) {
+        let lower = match self.lower {
+            Limit::Unbounded => 0,
+            Limit::Closed(x) => x,
+            Limit::Open(x) => x + 1,
+        };
+        let upper = match self.upper {
+            Limit::Unbounded => usize::MAX,
+            Limit::Closed(x) => x,
+            Limit::Open(x) => x - 1,
+        };
+        (lower, upper)
+    }
+}
+
+// TODO: consider changing to Range<NonZeroU64> since this is only used for the `precision` constraint
+pub type U64Range = Range<u64>;
```

Comment:
What are your thoughts on this TODO?

Reply:
I did not understand what its suggesting. But I think even if its only used by `precision` we should still keep it.

Reply:
The TODO is suggesting this:
```suggestion
pub type NonZeroU64Range = Range<NonZeroU64>;
```

Reply:
Oh yes, we should change it to `NonZeroU64`.